File: tests/leetcode/x/php/0087.php
<?php
function isScramble($s1, $s2) {
    $memo = [];
    $dfs = function($i1, $i2, $len) use ($s1, $s2, &$memo, &$dfs) {
        $key = $i1 . "," . $i2 . "," . $len;
        if (array_key_exists($key, $memo)) return $memo[$key];
        $a = substr($s1, $i1, $len);
        $b = substr($s2, $i2, $len);
        if ($a === $b) return $memo[$key] = true;
        $cnt = array_fill(0, 26, 0);
        for ($i = 0; $i < $len; $i++) {
            $cnt[ord($a[$i]) - 97]++;
            $cnt[ord($b[$i]) - 97]--;
        }
        foreach ($cnt as $v) if ($v != 0) return $memo[$key] = false;
        for ($k = 1; $k < $len; $k++) {
            if (($dfs($i1, $i2, $k) && $dfs($i1 + $k, $i2 + $k, $len - $k)) ||
                ($dfs($i1, $i2 + $len - $k, $k) && $dfs($i1 + $k, $i2, $len - $k))) return $memo[$key] = true;
        }
        return $memo[$key] = false;
    };
    return $dfs(0, 0, strlen($s1));
}

$lines = preg_split("/\\r?\\n/", stream_get_contents(STDIN));
if ($lines && trim($lines[0]) !== "") {
    $t = intval(trim($lines[0]));
    $out = [];
    for ($i = 0; $i < $t; $i++) $out[] = isScramble($lines[1 + 2 * $i], $lines[2 + 2 * $i]) ? "true" : "false";
    echo implode("\n", $out);
}
?>

File: tests/leetcode/x/erlang/0087.erl
#!/usr/bin/env escript
main(_) ->
    {ok, Input} = file:read_file("/dev/stdin"),
    Lines = string:split(binary_to_list(Input), "\n", all),
    case Lines of
        [] -> ok;
        [T0 | _] when T0 =:= "" -> ok;
        [T0 | Rest] ->
            T = list_to_integer(string:trim(T0)),
            io:put_chars(run(T, Rest, []))
    end.

run(0, _Rest, Out) -> string:join(lists:reverse(Out), "\n");
run(T, [S1, S2 | Rest], Out) ->
    M = maps:new(),
    Ans = case dfs(string:trim(S1), string:trim(S2), 1, 1, length(string:trim(S1)), M) of {true, _} -> "true"; _ -> "false" end,
    run(T - 1, Rest, [Ans | Out]).

dfs(S1, S2, I1, I2, Len, Memo) ->
    Key = {I1, I2, Len},
    case maps:find(Key, Memo) of
        {ok, V} -> {V, Memo};
        error ->
            A = string:substr(S1, I1, Len),
            B = string:substr(S2, I2, Len),
            case A =:= B of
                true -> {true, maps:put(Key, true, Memo)};
                false ->
                    case lists:sort(A) =:= lists:sort(B) of
                        false -> {false, maps:put(Key, false, Memo)};
                        true -> split_try(S1, S2, I1, I2, Len, 1, Memo, Key)
                    end
            end
    end.

split_try(_S1, _S2, _I1, _I2, Len, K, Memo, Key) when K >= Len -> {false, maps:put(Key, false, Memo)};
split_try(S1, S2, I1, I2, Len, K, Memo, Key) ->
    {A1, Memo1} = dfs(S1, S2, I1, I2, K, Memo),
    {A2, Memo2} = dfs(S1, S2, I1 + K, I2 + K, Len - K, Memo1),
    case A1 andalso A2 of
        true -> {true, maps:put(Key, true, Memo2)};
        false ->
            {B1, Memo3} = dfs(S1, S2, I1, I2 + Len - K, K, Memo2),
            {B2, Memo4} = dfs(S1, S2, I1 + K, I2, Len - K, Memo3),
            case B1 andalso B2 of
                true -> {true, maps:put(Key, true, Memo4)};
                false -> split_try(S1, S2, I1, I2, Len, K + 1, Memo4, Key)
            end
    end.
